pkg/controller/endpointslice/reconciler.go

```diff
@@ -0,0 +1,288 @@
+		}
+		if len(phSlicesToUpdate) > 0 {
+			slicesToUpdate = append(slicesToUpdate, phSlicesToUpdate...)
+		}
+		if phSliceNamesToDelete.Len() > 0 {
+			sliceNamesToDelete = sliceNamesToDelete.Union(phSliceNamesToDelete)
+		}
+	}
+
+	// if there are unique sets of ports that are no longer desired, mark
+	// the corresponding endpoint slices for deletion
+	portHashesToDelete := existingPortHashes.Difference(desiredPortHashes)
+	for portHashesToDelete.Len() > 0 {
+		portHash, _ := portHashesToDelete.PopAny()
+		matchingSlices := existingSlicesByPortHash[portHash]
+
+		for _, matchingSlice := range matchingSlices {
+			sliceNamesToDelete.Insert(matchingSlice.Name)
+		}
+	}
+
+	// when no endpoint slices would usually exist, we need to add a placeholder
+	if len(existingSlices) == sliceNamesToDelete.Len() && len(slicesToCreate) < 1 {
+		placeholderSlice := newEndpointSlice(r.service, EndpointMeta{Ports: []discovery.EndpointPort{}})
+		slicesToCreate = append(slicesToCreate, placeholderSlice)
+	}
+
+	// if there are slices to create and delete, change the creates to updates of the
+	// slices that would otherwise be deleted
+	for len(slicesToCreate) > 0 && sliceNamesToDelete.Len() > 0 {
+		sliceName, _ := sliceNamesToDelete.PopAny()
+		slice := slicesToCreate[len(slicesToCreate)-1]
+		slicesToCreate = slicesToCreate[:len(slicesToCreate)-1]
+		slice.Name = sliceName
+		slicesToUpdate = append(slicesToUpdate, slice)
+	}
+
+	for _, endpointSlice := range slicesToCreate {
+		addTriggerTimeAnnotation(endpointSlice, triggerTime)
+		_, err := r.client.DiscoveryV1alpha1().EndpointSlices(r.service.Namespace).Create(endpointSlice)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("Error creating EndpointSlice for Service %s/%s: %v", r.service.Namespace, r.service.Name, err))
+		}
+	}
+
+	for _, endpointSlice := range slicesToUpdate {
+		addTriggerTimeAnnotation(endpointSlice, triggerTime)
+		_, err := r.client.DiscoveryV1alpha1().EndpointSlices(r.service.Namespace).Update(endpointSlice)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("Error updating %s EndpointSlice for Service %s/%s: %v", endpointSlice.Name, r.service.Namespace, r.service.Name, err))
+		}
+	}
+
+	for sliceNamesToDelete.Len() > 0 {
+		sliceName, _ := sliceNamesToDelete.PopAny()
+		err := r.client.DiscoveryV1alpha1().EndpointSlices(r.service.Namespace).Delete(sliceName, &metav1.DeleteOptions{})
+		if err != nil {
+			errs = append(errs, fmt.Errorf("Error deleting %s EndpointSlice for Service %s/%s: %v", sliceName, r.service.Namespace, r.service.Name, err))
+		}
+	}
+
+	return utilerrors.NewAggregate(errs)
+}
+
+func (r *Reconciler) reconcileByPortHash(
+	existingSlices []discovery.EndpointSlice,
+	desiredEndpoints []discovery.Endpoint,
+	endpointMeta EndpointMeta,
+) ([]*discovery.EndpointSlice, []*discovery.EndpointSlice, sets.String) {
+	// build data structures for desired state
+	endpointsByHash := map[string]discovery.Endpoint{}
```

Comment:
Probably better to separate this into a util structure. Use this util structure to calculate desired ep list and current ep list. 
```

type EndpointSet struct {
endpointsByHash  map[string]discovery.Endpoint{}
hashes                  sets.String{}
}

func(*EndpointSet) Insert(discovery.Endpoint) ...
func(*EndpointSet) Delete(discovery.Endpoint) ...
func(*EndpointSet) List(discovery.Endpoint) ...
func(*EndpointSet) Has(discovery.Endpoint) ...

func(*EndpointSet) DIfference(discovery.Endpoint) ...// may not need this??
```

Reply:
this really helped clean things up, thanks!